lib/imagefs.coffee

```diff
@@ -46,17 +52,30 @@ driver = require('./driver')
 # 	stream.pipe(fs.createWriteStream('/bar/qux'))
 ###
 exports.read = (definition) ->
-	driver.interact(definition.image, definition.partition).then (fat) ->
-		return fat.createReadStream(definition.path)
-			.on('end', fat.closeDriver)
+	checkImageType(definition.image)
+	if _.isString(definition.image)
+		fs.openAsync(definition.image, 'r')
+		.then (fd) ->
+			close = -> fs.closeAsync(fd)
+			disk = new filedisk.FileDisk(fd, true)
+			read(disk, definition.partition, definition.path)
+			.tap (stream) ->
+				stream.on('end', close)
+				stream.on('error', close)
+	else if definition.image instanceof filedisk.Disk
+		read(definition.image, definition.partition, definition.path)
 
+read = (disk, partition, path) ->
```

Comment:
Done, I also moved all other helpers definitions above the functions that use them in this file.